py/dml/dmlparse.py

```diff
@@ -479,6 +479,32 @@ def maybe_default_no(t):
     '''maybe_default :'''
     t[0] = False
 
+
+def report_pretval(site, file_info, start, end, rparen, outp, stmts):
+    ends_with_return = stmts and stmts[-1].kind == 'return'
+    if (len(outp) == 1
+        and not ends_with_return
+        and stmts
+        and stmts[-1].kind == 'expression'
+        and stmts[-1].args[0].kind == 'set'):
+        (lh, rh) = stmts[-1].args[0].args
+        if lh.kind == 'variable_dml12' and lh.args[0] == outp[0][2]:
+            report(POUTARGRETURN(lh.site, start_site(rh.site), None))
+            # suppress PRETVAL's insertion of return statement
+            ends_with_return = True
+    report(PRETVAL(
+        site,
+        DumpableSite(file_info, start),
+        DumpableSite(file_info, end),
+        DumpableSite(file_info, rparen),
+        [(psite.loc(), pname) for (_, psite, pname, _) in outp]))
+    if not ends_with_return:
```

Comment:
This is overly aggressive, as method bodies may not be fallthrough (e.g. last statement may be `throw;`).

Reply:
True, but this is not a regression in this patch; previously ends_with_return was passed to PRETVAL, which triggered what is now PRETVAL_END; this is only a refactoring. I think bodies that end with throw give parse errors that are easy to fix.